Add cache in front of Dynamo lookup

samfrench opened the pull request:
Add cache in front of Dynamo lookup

https://jira.dev.bbc.co.uk/browse/CONNPOL-2821

![](https://media.giphy.com/media/V0ZyjR3QZkLyo/giphy.gif)

samfrench commented on lib/alephant/lookup.rb:
```diff
@@ -9,9 +9,9 @@ module Lookup
     @@lookup_tables = {}
 
-    def self.create(table_name)
+    def self.create(table_name, config={})
```
I am thinking of passing in Broker config here. I could pass in settings needed for the cache, but that might become quite a long constructor. I have the options of:
1) Keep as above.
2) Pass in table_name in config too, as it's in there as well.
3) Pass in params separately.
4) Somehow read the Broker config in this gem. Not sure if that is possible.

Integralist replied:
@samfrench I think the above is fine. I'm not really a big fan of method signatures that have a single argument which is just a hash of stuff. I like the idea that the method has a baseline requirement of one argument (the table name to be created) and then anything beyond that is configurable by passing in data via a config object

So I'd suggest leaving the above 'as is'. 

Reading in config from the Broker would be tightly coupling this gem to the broker and would cause other problems in the future I fear. So best to inject rather than couple the code

samfrench commented on lib/alephant/lookup/lookup_cache.rb:
```diff
@@ -0,0 +1,67 @@
+require 'dalli-elasticache'
```
I copied over this cache class from the Alephant-Broker repo and made some modifications.
> https://github.com/BBC-News/alephant-broker/blob/master/lib/alephant/broker/cache.rb

Integralist replied:
@samfrench I think I should probably create a new gem of reusable 'modules' that can be `include`'d into different classes.

That way we can do simple things like changing:

```rb
logger.debug "Alephant::LookupCache::#initialize: No config endpoint, NullClient used"
```

to

```rb
logger.debug "#{self.class.name}##{__method__}: No config endpoint, NullClient used"
```

But also try and find clever ways of injecting the differences between two separate gems when including the same module 

e.g. your new `cache.rb` vs the brokers `cache.rb` is subtly different, but very much the same so there is a lot of duplication there and it'll get worse if we need to replicate the `cache.rb` from this PR into the Sequencer as well. So if I moved that code into a reusable 'cache' module, I might be able to then inject the differences via a lambda or some other such mechanism.

I'll discuss with @jfro5 about this. It's likely that this PR stays as you've done it (for now) and I'll get a ticket created for the work to be cleaned up via this new gem.

samfrench commented on lib/alephant/lookup/lookup_helper.rb:
```diff
@@ -21,15 +23,17 @@ def initialize(lookup_table)
-            "batchVersion" => batch_version,
-            "method"       => "#{self.class}#read"
-          )
+        LookupCache.new(@config).get(component_cache_key(id, opts, batch_version)) do
```
I placed the cache in the lookup helper. I wasn't sure of the best place to add it, but thought the query was too low down. I am hoping to store and retrieve a LookupLocation object from the cache, so it is consistent with the object LookupQuery returns, and also the same return type as the write method.

samfrench commented on spec/lookup_spec.rb:
```diff
@@ -32,24 +30,28 @@
             "component_key"=> {
               :comparison_operator=>"EQ",
-              :attribute_value_list=>[{"s"=>"id/9dd916afd5516828a91d259967fd394a"}]
+              :attribute_value_list=>[{"s"=>"id/218c835cec343537589dbf1619532e4d"}]
```
I swapped the order of the params around, with the options and the batch_version in these tests, as I noticed the instance being tested has them the other way around.

samfrench commented on spec/lookup_spec.rb:
```diff
@@ -32,24 +30,28 @@
-          .stub(:initialize)
+      it "queries DynamoDb and returns a location when not in cache" do
+        expect_any_instance_of(Dalli::ElastiCache).to receive(:initialize)
+        expect_any_instance_of(Dalli::ElastiCache).to receive(:client).and_return(Dalli::Client.new)
```
Is it best to return a new object (which would be the object the class would return), or a double? I kept changing it, and decided to leave it as this for now.

Also the `expect_any_instance_of` somehow leaks out into the next test! I'll wait for feedback on the above and revisit if it's needed. I took it out of the next test for now, so they all pass. :wink:

Integralist replied:
@samfrench re: returning an actual object instance vs returning a double. I'd personally stick with returning an actual instance unless I needed to mock/stub any methods on the object. So keep it as is for now.

samfrench commented:
@BBCMarkMcDonnell, here is my attempt at adding a cache to the Lookup. This is my first non-minor Alephant change, so don't be too harsh on the code! I look forward to the feedback. The Sequencer one will definitely be easier, once this one has been reviewed and changed.

samfrench commented:
One last point, this will probably take 30+ minutes to review.

Integralist commented on lib/alephant/lookup/lookup_helper.rb:
```diff
@@ -55,6 +59,14 @@ def write(id, opts, batch_version, location)
+      private
+
+      def component_cache_key(id, opts, batch_version)
+        "#{@lookup_table.table_name}/COMPONENT_KEY/#{batch_version}".gsub("COMPONENT_KEY") do |s|
```
@samfrench can change this to something like:

```rb
def component_cache_key(id, opts, batch_version)
  template_key(batch_version).gsub("{{component_key}}") do |s|
    LookupLocation.new(id, opts, batch_version).component_key
  end
end

def template_key(batch_version)
  "#{@lookup_table.table_name}/{{component_key}}/#{batch_version}"
end
```

Just felt there was too much going on in that one method. Also I prefer seeing a more typical template syntax (e.g. Mustache) used when doing interpolation; for consistency reasons more than anything else as that's a well known syntax for template interpolation. I'm sure this would be a contentious argument but that's my opinion on it :-)

Integralist commented on lib/alephant/lookup/lookup_cache.rb:
```diff
@@ -0,0 +1,67 @@
+      DEFAULT_TTL  = 5
+
+      def initialize(config)
+        @config = config || {}
```
@samfrench Is there a reason you set the default value inside the method rather than in the method signature? e.g. `initialize(config={})`

There isn't a wrong or right answer per se. The way it is at the moment means that a config *must* be provided when creating a new instance of the class, where as you can make the class a little less fraile by having the argument default value defined within the signature.

Integralist replied:
@samfrench I would add an `attr_reader :config` and then modify the latter calls to `@config` to just `config`. Reads a bit nicer

Integralist commented on lib/alephant/lookup/lookup_helper.rb:
```diff
@@ -10,8 +11,9 @@ class LookupHelper
-      def initialize(lookup_table)
+      def initialize(lookup_table, config)
         @lookup_table = lookup_table
+        @config = config
```
I would add an `attr_reader :config` and then modify any latter calls to `@config` to `config`. Nicer to read.

Ps, you've referenced `@lookup_table` in some places and that already has a reader method so can you change those to be just `lookup_table`. Again, I just feel it looks nicer

Integralist commented on spec/lookup_spec.rb:
```diff
@@ -17,8 +15,8 @@
     describe "#initialize(table_name)" do
```
@samfrench I would look to modify this (and any other tests you notice) so that their `describe` doesn't include the specific signature. Because as you can see, in this case the signature has changed and the spec is reading incorrectly. I'd prefer to just have it as `describe "#initialize"` so that it can adapt as the code requirements change.

Integralist commented on spec/lookup_spec.rb:
```diff
@@ -17,8 +15,8 @@
-        table.should_receive(:table_name)
-        subject.new(table)
+        expect(table).to receive(:table_name)
+        subject.new(table, nil)
```
@samfrench I'd also probably just call `subject.new(table)` and not bother with passing `nil` as the second argument.

Integralist commented:
Hi @samfrench left some comments for you. Thanks

samfrench commented:
@BBCMarkMcDonnell, I finally addressed all of the points you raised/suggested.

* It is a good suggestion about a reusable gem. Obviously that will be sometime, and the framework may change in other ways, so we will go with this for now.
* I moved the initialising default value to the constructor as I wasn't doing it underneath for any particular reason. I thought nil could be passed in if other gems weren't passing in the second parameter, then I thought nil would override the default constructor?
* I agree with the point about using mustache syntax for interpolation. I used to use it, but other people advise against it. It's just personal preference, and I do prefer it.
* I removed the method signatures from the rspec describe blocks, as they can get out of date. Looking at previous code written, I don't usually include them.
* I haven't sorted out the test leaking into the next, as I'm not sure if that's a must-have for this PR. I noticed previously this was an issue, so it is still present. I could raise a ticket for that, or that could halt this PR until resolved.

Any other points for this PR, or is this ready to go?

BBCMarkMcDonnell commented:
Hi Sam, no problem. Just to let you know that I'll review this later today.

Integralist commented:
Hi @samfrench the test can be deferred to another PR if you've not got time right now. As long as the tests are passing then their leakiness can be cleaned up later

Integralist commented:
Otherwise :cake:

BBCMarkMcDonnell commented:
Hi @samfrench 

Using an instance variable for the Dalli client *should* be OK in the sense that it ultimately points to a singular ElastiCache endpoint. so if there was a concurrency issue whereby multiple clients were created, they would still ultimately be calling the same underlying cache service (as far as I'm aware - don't know if @zygotic99 would argue that's bad logic on my part)

zygotic99 commented:
@Integralist / @BBCMarkMcDonnell I think that all makes sense.  That's how Newsbeat uses Elasticache - the same underlying service endpoint - and although Newsbeat isn't a paragon of good code it does work and, when fixed in AMP, the Elasticache code does work.  Erm...

Anyway, sounds fine to me  

@samfrench
